pulsar-broker/src/main/java/org/apache/pulsar/compaction/CompactedTopicUtils.java

```diff
@@ -88,15 +92,27 @@ public static void asyncReadCompactedEntries(TopicCompactionService topicCompact
                             return;
                         }
 
+                        List<Entry> unAckedEntries = new ArrayList<>();
                         long entriesSize = 0;
-                        for (Entry entry : entries) {
-                            entriesSize += entry.getLength();
+                        if (COMPACTION_CURSOR_NAME.equals(cursor.getName())) {
+                            for (Entry entry : entries) {
+                                entriesSize += entry.getLength();
+                            }
+                            unAckedEntries = entries;
+                        } else {
+                            for (Entry entry : entries) {
+                                Position position = entry.getPosition();
+                                if (!cursor.isMessageIndividualDeleted(position)) {
+                                    unAckedEntries.add(entry);
+                                    entriesSize += entry.getLength();
+                                }
+                            }
                         }
-                        cursor.updateReadStats(entries.size(), entriesSize);
+                        cursor.updateReadStats(unAckedEntries.size(), entriesSize);
 
                         Entry lastEntry = entries.get(entries.size() - 1);
                         cursor.seek(lastEntry.getPosition().getNext(), true);
-                        callback.readEntriesComplete(entries, readEntriesCtx);
+                        callback.readEntriesComplete(unAckedEntries, readEntriesCtx);
```

Comment:
This will cause these messages to be lost the next time compaction is run, and other consumers will not be able to read these messages. so why use individual ack for Exclusive mode?

Reply:
> for the compaction_cursor_name __compaction , we don't skip the individual acked position, this will not affect the compaction

OK， I see

> The individual ack is not recommended in Exclusive mode? I don't know about this, I always use this before😂

Yes, for Exclusive mode, suggest using `acknowledgeCumulative` acknowledge. actually, if every message is confirmed, then acknowledgment and acknowledgment have the same effect, for this cause, I think #21187 has fixed it. But if the acknowledged messages are discontinuous then we can receive acknowledged messages, for this case, I think is a misuse of individual acknowledge in Exclusive mode

Reply:
@summeriiii I think for the compacted topic, we cannot use individual acknowledgment, otherwise the entry will not be deleted because the messages deleted by compact will never be acknowledged. so I suggest using `acknowledgeCumulative` acknowledge message for compacted topic.